File: src/main/java/no/novari/configuration/validation/valueparsability/validators/ValueConvertingParsabilityValidator.java
package no.novari.configuration.validation.valueparsability.validators;

import no.novari.configuration.model.configuration.entities.ValueMapping;
import no.novari.configuration.validation.valueparsability.ValueParsabilityValidator;
import org.springframework.stereotype.Service;

import java.util.regex.Pattern;

@Service
public class ValueConvertingParsabilityValidator implements ValueParsabilityValidator {

    private static final Pattern numberPattern = Pattern.compile("\\d+");
    private static final Pattern valueConverterReferencePattern = Pattern.compile("\\$vc\\{" + numberPattern + "}");

    private static final Pattern instanceValueKeyPattern = Pattern.compile("(?:(?!\\$if\\{).)+");
    private static final Pattern ifReferencePattern = Pattern.compile("\\$if\\{" + instanceValueKeyPattern + "}");
    private static final Pattern instanceCollectionFieldReferencePattern = Pattern.compile("\\$icf\\{" + numberPattern + "}\\{" + instanceValueKeyPattern + "}");


    private static final Pattern valueConvertingPattern = Pattern.compile("^" + valueConverterReferencePattern + "(" + ifReferencePattern + "|" + instanceCollectionFieldReferencePattern + ")$");

    @Override
    public ValueMapping.Type getTypeToValidate() {
        return ValueMapping.Type.VALUE_CONVERTING;
    }

    @Override
    public boolean isValid(String value) {
        return valueConvertingPattern.matcher(value).matches();
    }

}

File: src/main/kotlin/no/novari/flyt/configuration/validation/valueparsability/validators/ValueConvertingParsabilityValidator.kt
package no.novari.flyt.configuration.validation.valueparsability.validators

import no.novari.flyt.configuration.model.configuration.entities.ValueMapping
import no.novari.flyt.configuration.validation.valueparsability.ValueParsabilityValidator
import org.springframework.stereotype.Service
import java.util.regex.Pattern

@Service
class ValueConvertingParsabilityValidator : ValueParsabilityValidator {
    private val numberPattern = Pattern.compile("\\d+")
    private val valueConverterReferencePattern =
        Pattern.compile(
            "${Pattern.quote(VALUE_CONVERTER_PREFIX)}$numberPattern\\}",
        )

    private val instanceValueKeyPattern = Pattern.compile("(?:(?!${Pattern.quote(IF_REFERENCE_PREFIX)}).)+")
    private val ifReferencePattern = Pattern.compile("${Pattern.quote(IF_REFERENCE_PREFIX)}$instanceValueKeyPattern\\}")
    private val instanceCollectionFieldReferencePattern =
        Pattern.compile(
            "${Pattern.quote(INSTANCE_COLLECTION_FIELD_PREFIX)}$numberPattern\\}\\{$instanceValueKeyPattern\\}",
        )

    private val valueConvertingPattern =
        Pattern.compile(
            "^$valueConverterReferencePattern($ifReferencePattern|$instanceCollectionFieldReferencePattern)$",
        )

    override fun getTypeToValidate(): ValueMapping.Type = ValueMapping.Type.VALUE_CONVERTING

    override fun isValid(value: String): Boolean = valueConvertingPattern.matcher(value).matches()

    private companion object {
        const val VALUE_CONVERTER_PREFIX = "\$vc{"
        const val IF_REFERENCE_PREFIX = "\$if{"
        const val INSTANCE_COLLECTION_FIELD_PREFIX = "\$icf{"
    }
}
